### PR TITLE
Use correct layouts when ad-free for `dynamic/slow-mpu`

### PR DESCRIPTION
**Pull request by OllysCoding**
Closes https://github.com/guardian/dotcom-rendering/issues/8053

Fixes an issue where DCR was incorrectly rendering layouts for dynamic/slow-mpu when the user was ad free.

This ends up being quite a big refactor, but by using the 'two slice' layout like we do in DynamicSlow & DynamicFast components - we're able to more easily support the ad-free layouts, which only affect the second slice.

The logic for these different layouts in Frontend can be found [here](https://github.com/guardian/frontend/blob/main/common/app/layout/slices/DynamicSlowMpu.scala#L25)

I've added some stories which cover the different variations of the 'ad-free' layouts depending on whether there's a first slice & the number of cards. There could be a few edge cases though that aren't covered, since Frontend's CSS does a lot of stretching & changing when the number of cards aren't what is expected. From a quick look around the fronts I don't see any of these crop up though.

## Screenshots

| Before      | After      | Frontend      |
| ----------- | ---------- | ---------- |
| ![before][] | ![after][] | ![frontend][]  |

[before]: https://github.com/guardian/dotcom-rendering/assets/9575458/4707151c-d623-45a3-a52f-50c444bc8d36
[after]: https://github.com/guardian/dotcom-rendering/assets/9575458/79294af5-342e-4327-beff-fd117314d2f6
[frontend]: https://github.com/guardian/dotcom-rendering/assets/9575458/54c37bff-6957-4df5-bd6f-daf9b97e300b

### CODE REVIEW

File: dotcom-rendering/src/components/DynamicSlowMPU.tsx [hunk 1]
```diff
@@ -128,216 +156,206 @@ const ColumnOfThree50_Ad50 = ({
...
-	adIndex: number;
-};
-
-const MPUSlice = ({
+/**
+ * DynamicSlowMPU
+ *
+ * This container only allows big and standard cards (from groupedTrails)
+ *
+ * Layout is dynamic deopending on the number of big cards. You're only
```
Comment (ioannakok) [inline dotcom-rendering/src/components/DynamicSlowMPU.tsx:164]: 👩‍🏫 

```suggestion
 * Layout is dynamic depending on the number of big cards. You're only
```

File: dotcom-rendering/src/components/DynamicSlowMPU.tsx [hunk 2]
```diff
@@ -128,216 +156,206 @@ const ColumnOfThree50_Ad50 = ({
...
-		case 1: {
-			// If there is only one big item a standard item is promoted such that
-			// there is always at least two
-			const promoted = groupedTrails.standard.slice(0, 1);
-			bigCards = [...groupedTrails.big, ...promoted];
-			standardCards = groupedTrails.standard.slice(1);
-			layout = 'twoBigs';
-			break;
+	let firstSliceCards: DCRFrontCard[] = [];
+	if (groupedTrails.big.length === 1) {
```
Comment (ioannakok) [inline dotcom-rendering/src/components/DynamicSlowMPU.tsx:187]: What happens if we receive 1 big and boosted? Not sure if that's even possible!

File: dotcom-rendering/src/components/DynamicSlowMPU.tsx [hunk 3]
```diff
@@ -128,216 +156,206 @@ const ColumnOfThree50_Ad50 = ({
...
+		// If there is only one big item a standard item is promoted such that
+		// there is always at least two
+		const promoted = groupedTrails.standard.slice(0, 1);
+		firstSliceCards = [...groupedTrails.big, ...promoted];
+		firstSliceLayout = 'twoBigs';
```
Comment (ioannakok) [inline dotcom-rendering/src/components/DynamicSlowMPU.tsx:192]: What happens if we receive 1 big and boosted? Should the layout become `twoBigsFirstBoosted`. Not sure if that's even possible, just asking!

File: dotcom-rendering/src/components/DynamicSlowMPU.tsx [hunk 4]
```diff
@@ -128,216 +156,206 @@ const ColumnOfThree50_Ad50 = ({
...
-		case 2: {
-			bigCards = groupedTrails.big;
-			standardCards = groupedTrails.standard;
-			if (groupedTrails.big[0]?.isBoosted) {
-				layout = 'twoBigsFirstBoosted';
-			} else if (groupedTrails.big[1]?.isBoosted) {
-				layout = 'twoBigsSecondBoosted';
+	} else if (groupedTrails.big.length > 2) {
+		firstSliceCards = groupedTrails.big.slice(0, 3);
+		firstSliceLayout = 'threeBigs';
```
Comment (ioannakok) [inline dotcom-rendering/src/components/DynamicSlowMPU.tsx:204]: We don't care about boosting when there are three big cards?

File: dotcom-rendering/src/components/DynamicSlowMPU.tsx [hunk 5]
```diff
@@ -128,216 +156,206 @@ const ColumnOfThree50_Ad50 = ({
...
+		filter: firstSliceCards,
+	});
+
+	let secondSliceLayout:
+		| 'noFirstSlice'
+		| 'standard'
+		| 'noFirstSliceNoMPU'
+		| 'noFirstSliceNoMPUThreeOrMore'
+		| 'standardNoMPU'
+		| 'standardNoMPUFourOrMore';
```
Comment (ioannakok) [inline dotcom-rendering/src/components/DynamicSlowMPU.tsx:220]: This is just personal preference so feel free to ignore but maybe we could rename `standard` -> `standardWithMPU`. I got a bit confused why we were only using `NoMPU`.

File: dotcom-rendering/src/components/DynamicSlowMPU.tsx (hunk 3 above)
Comment (ioannakok) [reply dotcom-rendering/src/components/DynamicSlowMPU.tsx:192]: Ignore this, I hadn't had coffee ☕

File: dotcom-rendering/src/components/DynamicSlowMPU.tsx (hunk 4 above)
Comment (OllysCoding) [reply dotcom-rendering/src/components/DynamicSlowMPU.tsx:204]: As fas as I can tell - no, this was the existing logic before the refactor too

File: dotcom-rendering/src/components/DynamicSlowMPU.tsx (hunk 5 above)
Comment (OllysCoding) [reply dotcom-rendering/src/components/DynamicSlowMPU.tsx:220]: I actually had it like this before but changed it, the 'default' state of the container is with an MPU, so I thought it was over verbose - think I will leave it as-is for now!

File: dotcom-rendering/src/components/DynamicSlowMPU.tsx [hunk 6]
```diff
@@ -128,216 +156,206 @@ const ColumnOfThree50_Ad50 = ({
...
+ * Layout is dynamic deopending on the number of big cards. You're only
+ * allowed to have 2 or 3 big cards. If you pass 1 a standard card will
+ * get promoted to make two bigs. If you pass more than 3 bigs then the
+ * extras will get demoted to standard.
+ *
+ * If you pass no bigs at all the top slice will not render and a special
+ * 3 column layout is used for the remaining slice.
+ *
+ */
+export const DynamicSlowMPU = ({
```
[the remark this answers is not in the document]
Comment (OllysCoding) [reply dotcom-rendering/src/components/DynamicSlowMPU.tsx:173]: It's a very tricky component (as are dynamic fast, slow & package) and we (at least) have an established practice for how they're written, but they do remain large & logic heavy!

**Review by ioannakok — APPROVED** [review]
Great work as usual!

**Pull request merged**